src/controllers/krate/versions.rs

```diff
@@ -21,27 +27,217 @@ pub async fn versions(state: AppState, Path(crate_name): Path<String>) -> AppRes
+                    .or(versions::created_at.lt(created_at)),
+            )
+        }
+        query = query.limit(options.per_page);
+    }
+
+    query = query.order((versions::created_at.desc(), versions::id.desc()));
+
+    let data: Vec<(Version, Option<User>)> = query.load(conn)?;
+    let mut next_page = None;
+    if let Some(options) = options {
+        next_page = next_seek_params(&data, options, |last| {
+            seek::Date(last.0.created_at, last.0.id)
+        })?
+        .map(|p| req.query_with_params(p));
+    };
+
+    // Since the total count is retrieved through an additional query, to maintain consistency
+    // with other pagination methods, we only make a count query while data is not empty.
+    let total = if !data.is_empty() {
+        versions::table.count().get_result(conn)?
+    } else {
+        0
+    };
+
+    Ok(PaginatedVersionsAndPublishers {
+        data,
+        meta: ResponseMeta { total, next_page },
+    })
+}
+
+// Unfortunately, Heroku Postgres has no support for the semver PG extension.
+// Therefore, we need to perform both sorting and pagination manually on the server.
+fn list_by_semver(
+    krate: &Crate,
+    options: Option<&PaginationOptions>,
+    req: &Parts,
+    conn: &mut PgConnection,
+) -> AppResult<PaginatedVersionsAndPublishers> {
+    let (data, total) = if let Some(options) = options {
+        // Sorting by semver but opted for id as the seek key because num can be quite lengthy,
+        // while id values are significantly smaller.
+        let mut sorted_versions = IndexMap::new();
+        for result in krate
+            .all_versions()
+            .select((versions::id, versions::num))
+            .load_iter::<(i32, String), DefaultLoadingMode>(conn)?
+        {
+            let (id, num) = result?;
+            sorted_versions.insert(id, (num, None));
+        }
+        sorted_versions.sort_by_cached_key(|_, (num, _)| Reverse(semver::Version::parse(num).ok()));
+
+        let mut idx = Some(0);
+        if let Page::Seek(ref seek) = options.page {
+            idx = seek
+                .decode::<i32>()
+                .map(|id| sorted_versions.get_index_of(&id))?
+                .filter(|i| i + 1 < sorted_versions.len())
+                .map(|i| i + 1);
+        }
+        if let Some(start) = idx {
+            let end = (start + options.per_page as usize).min(sorted_versions.len());
+            let ids = sorted_versions[start..end].keys().collect::<Vec<_>>();
+            for result in krate
+                .all_versions()
+                .left_outer_join(users::table)
+                .select((versions::all_columns, users::all_columns.nullable()))
+                .filter(versions::id.eq_any(ids))
+                .load_iter::<(Version, Option<User>), DefaultLoadingMode>(conn)?
+            {
+                let row = result?;
+                sorted_versions.insert(row.0.id, (row.0.num.to_owned(), Some(row)));
```

Comment:
it might be worth adding a comment somewhere in this function (maybe on the `let sorted_versions`) explaining the purpose of that map and why we're reusing the `sorted_versions` map here. the naive implementation would collect the results in a `Vec` and sort that again, but I assume the reason for reusing `sorted_versions` is to not run `Version::parse()` again?

Reply:
Yes, kind of. We can either fetch all joined data at once and perform sorting and pagination on it, or fetch only the data needed for sorting and pagination, then make another query for the sorted and paginated response.

Due to the fact that the versions would only increase in the future and sorting and pagination need to happen on the app server, I think the second approach, which minimizes payload and memory usage, would be better. In this case, we can utilize the sorted map and enrich it later without sorting twice.